### PR TITLE
[dagster-dbt] Allow SDAs generated with load_assets_from_dbt* to be partitioned

### PR DESCRIPTION
**Pull request by OwenKephart**
### Summary & Motivation

Addresses: https://github.com/dagster-io/dagster/issues/7683

A very minimal implementation allowing for partitioned dbt assets. Adds two new parameters, partitions_def and partition_key_to_vars_fn, which allows the user to define a function that takes the current partition key (e.g. "2022-01-01") and returns a dictionary (e.g. `{"run_date": "2022-01-01"}`).

Users that want to have multiple different partitions will need to load_assets_from_dbt_... multiple times, e.g.

```python

hourly_dbt_assets = load_assets_from_dbt_project(..., select="tag:hourly", partitions_def=HourlyPartitionsDefinition(...))
daily_dbt_assets = load_assets_from_dbt_project(..., select="tag:daily", partitions_def=DailyPartitionsDefinition(...))
```

### CODE REVIEW

**Comment by sryza** [comment]
Pretty straightforward!

**Comment by sryza** [comment]
I can't think of any great way of dealing it, but something weird here that I think is worth pointing out is the difference in semantics between:
* How our partitioned IO managers work (e.g. the snowflake_io_manager) - the entire partition is dropped and then re-written.
* How people are be encouraged to do things in dbt - overwriting by unique ID.

File: python_modules/libraries/dagster-dbt/dagster_dbt/asset_defs.py [hunk 1]
```diff
@@ -318,6 +327,8 @@ def load_assets_from_dbt_project(
     io_manager_key: Optional[str] = None,
     node_info_to_asset_key: Callable[[Mapping[str, Any]], AssetKey] = _get_node_asset_key,
     use_build_command: bool = False,
+    partitions_def: Optional[PartitionsDefinition] = None,
+    partition_key_to_vars_fn: Optional[Callable[[str], Dict[str, Any]]] = None,
```
[the remark this answers is not in the document]
Comment (OwenKephart) [reply python_modules/libraries/dagster-dbt/dagster_dbt/asset_defs.py:331]: probably would just change the signature of this function to accept either a string or list of strings

**Comment by sryza** [comment]
> should, in almost all cases, be using either the delete+insert or insert_overwrite strategy for their incremental models, rather than merge.

My read is that the semantics of "delete+insert" are exactly the same as the semantics of "merge", and neither of these match what the snowflake IO manager does.

Imagine you have a daily-partitioned model called "filtered_events", where each event has a timestamp, and you want to be able to overwrite day partitions. What would you use as your dbt unique_key?

**Comment by OwenKephart** [comment]
> Imagine you have a daily-partitioned model called "filtered_events", where each event has a timestamp, and you want to be able to overwrite day partitions. What would you use as your dbt unique_key?

It's pretty difficult to find actual documentation for delete+insert, but my understanding is that it takes the unique_key and deletes any existing rows who have a unique_key that matches a row in the new data that dbt wants to add in. This means that you can have a run_date column, and if you run dbt for a run_date=2020-01-01, then all of your new data should have some column called run_date, where the value is 2020-01-01. Then, dbt will delete any existing data that matches that run date (so it will get rid of all the old 2020-01-01 data), and finally insert all the new data. These are the semantics that we'd expect.

This is different from the original merge strategy, because the merge strategy requires that each of the values in the unique_key column are actually unique (whereas you'll have a huge number of duplicates if you use run_date as a unique_key)

**Comment by sryza** [comment]
> It's pretty difficult to find actual documentation for delete+insert, but my understanding is that it takes the unique_key and deletes any existing rows who have a unique_key that matches a row in the new data that dbt wants to add in.

Ah, interesting

File: python_modules/libraries/dagster-dbt/dagster_dbt/asset_defs.py [hunk 2]
```diff
@@ -389,6 +408,8 @@ def load_assets_from_dbt_manifest(
     selected_unique_ids: Optional[AbstractSet[str]] = None,
     node_info_to_asset_key: Callable[[Mapping[str, Any]], AssetKey] = _get_node_asset_key,
     use_build_command: bool = False,
+    partitions_def: Optional[PartitionsDefinition] = None,
+    partition_key_to_vars_fn: Optional[Callable[[str], Dict[str, Any]]] = None,
```
Comment (sryza) [inline python_modules/libraries/dagster-dbt/dagster_dbt/asset_defs.py:412]: Nitpick: Mapping instead of Dict

**Review by sryza — APPROVED** [review]
Nice

**Pull request merged**